packages/react-combobox/src/components/Combobox/useCombobox.ts

```diff
@@ -11,18 +23,167 @@ import type { ComboboxProps, ComboboxState } from './Combobox.types';
  * @param props - props from this instance of Combobox
  * @param ref - reference to root HTMLElement of Combobox
  */
-export const useCombobox_unstable = (props: ComboboxProps, ref: React.Ref<HTMLElement>): ComboboxState => {
+export const useCombobox_unstable = (
+  props: ComboboxProps,
+  optionCollection: OptionCollectionState,
+  ref: React.Ref<HTMLButtonElement>,
+): ComboboxState => {
+  const {
+    inline = false,
+    multiselect,
+    onOpenChange,
+    open: controlledOpen,
+    placeholder,
+    positioning,
+    value: controlledValue,
+  } = props;
+  const {
+    options,
+    collectionData: { count, getOptionAtIndex, getIndexOfKey, getOptionByKey },
+  } = optionCollection;
+  const idBase = useId('combobox');
+
+  const [activeOption, setActiveOption] = React.useState<OptionValue | undefined>();
+  const [selectedKeys, selectKey] = useSelection(props);
+
+  const [value, setValue] = useControllableState({
+    state: controlledValue,
+    initialState: undefined,
+  });
+
+  const [open, setOpenState] = useControllableState({
+    state: controlledOpen,
+    initialState: false,
+  });
+
+  // popper
+  const popperOptions = {
+    enabled: open,
+    position: 'below' as const,
+    align: 'start' as const,
+    ...resolvePositioningShorthand(positioning),
+  };
+
+  const {
+    targetRef: popperTargetRef,
+    containerRef: popperContainerRef,
+  }: {
+    targetRef: React.MutableRefObject<HTMLButtonElement>;
+    containerRef: React.MutableRefObject<HTMLDivElement>;
+  } = usePopper(popperOptions);
+
+  // update value based on selectedKeys
+  React.useEffect(() => {
+    let newValue;
+    if (multiselect) {
+      newValue = selectedKeys.map(key => getOptionByKey(key).value).join(', ');
+    } else {
+      const selectedOption = getOptionByKey(selectedKeys[0]);
+      newValue = selectedOption ? selectedOption.value : placeholder;
+    }
+
+    setValue(newValue);
+  }, [getOptionByKey, multiselect, placeholder, selectedKeys, setValue]);
+
+  const setOpen = (event: OpenEvents, newState: boolean) => {
+    onOpenChange?.(event, { open: newState });
+    setOpenState(newState);
+  };
+
+  const onOptionClick = (event: React.MouseEvent<HTMLElement>, optionKey: string) => {
```

Comment:
Since you already have a context for the component, it could be a cleaner implem pass handlers like `setActiveOption` and `setOpen` through context to the options who can control this.  I think it will result in better encapsulation and delegates some complexity/responsibility to the options.

Reply:
I actually initially coded it that way, and decided it was less clean than passing through `onOptionClick`. There are essentially two reasons:
1. Listbox and Combobox have slightly different logic for whether `setOpen` should be called on click
2. It's actually easier to reason about the overall activeOption and selection logic if the click handler and keyboard enter/space handler are in the same place. Otherwise you end up needing to jump back and forth between the `useCombobox`/`useListbox` and `useOption` files, `useOption` needs to have split logic for Listbox vs. Combobox, and it all gets a bit messy.